File: backend/state/models/character_profile.py
from dataclasses import dataclass, fields


@dataclass
class CharacterProfile:
    species: str = ""
    background: str = ""
    alignment: str = ""
    pronouns: str = ""
    age: str = ""
    height: str = ""
    weight: str = ""
    eyes: str = ""
    skin: str = ""
    hair: str = ""
    appearance: str = ""
    personality_traits: str = ""
    ideals: str = ""
    bonds: str = ""
    flaws: str = ""
    allies_and_organizations: str = ""
    backstory: str = ""

    def __post_init__(self) -> None:
        for profile_field in fields(self):
            value = getattr(self, profile_field.name)
            if not isinstance(value, str):
                raise ValueError(
                    f"Character profile field '{profile_field.name}' must be text."
                )
            setattr(self, profile_field.name, value.strip())

    @classmethod
    def from_dict(cls, raw: dict | None) -> "CharacterProfile":
        if raw is None:
            return cls()
        if not isinstance(raw, dict):
            raise ValueError("Character profile must be an object.")
        return cls(
            **{
                profile_field.name: raw.get(profile_field.name, "")
                for profile_field in fields(cls)
            }
        )

File: frontend/src/features/sheets/characterProfile.ts
import type { CharacterProfile } from "@/domain/models";

export const CHARACTER_PROFILE_FIELDS = [
  "species",
  "background",
  "alignment",
  "pronouns",
  "age",
  "height",
  "weight",
  "eyes",
  "skin",
  "hair",
  "appearance",
  "personality_traits",
  "ideals",
  "bonds",
  "flaws",
  "allies_and_organizations",
  "backstory"
] as const satisfies readonly (keyof CharacterProfile)[];

export function createEmptyCharacterProfile(): CharacterProfile {
  return Object.fromEntries(
    CHARACTER_PROFILE_FIELDS.map((field) => [field, ""])
  ) as unknown as CharacterProfile;
}

export function resolveCharacterProfile(
  profile: Partial<CharacterProfile> | null | undefined
): CharacterProfile {
  const resolved = createEmptyCharacterProfile();
  for (const field of CHARACTER_PROFILE_FIELDS) {
    resolved[field] = profile?.[field] ?? "";
  }
  return resolved;
}

export function normalizeCharacterProfile(profile: CharacterProfile): CharacterProfile {
  return Object.fromEntries(
    CHARACTER_PROFILE_FIELDS.map((field) => [field, profile[field].trim()])
  ) as unknown as CharacterProfile;
}
